### ISSUE TITLE
Circular reference between tsconfig files leads to deadlock

### ISSUE THREAD

**Issue by bitschubse** [aleclarson/vite-tsconfig-paths#132]
After updating to version 4.3.1 (same problem for 4.3.0), our build hangs directly after the build command call. There are no log messages or errors.

Using the DEBUG variable we get only the following output. Build hangs after that.
```
  vite-tsconfig-paths options.root   == /home/user/app +0ms
  vite-tsconfig-paths project root   == /home/user/app +1ms
  vite-tsconfig-paths workspace root == undefined +0ms
  vite-tsconfig-paths projects: [
  '/home/user/app/tsconfig.json',
  '/home/user/app/project1/tsconfig.json',
  '/home/user/app/project2/tsconfig.json',
  '/home/user/app/project3/tsconfig.json'
] +71ms
```

Used vite dependency versions:
```
"vite": "4.5.1",
"vite-plugin-dts": "3.7.0",
"vite-plugin-eslint": "1.8.1",
"vite-tsconfig-paths": "4.3.1",
```

Does anyone have any idea what the cause is or how we can get to the bottom of the problem?

**Comment by bitschubse** [comment]
Unfortunately, it is not so easy to reduce our project to a easy StackBlitz reproduction. Therefore I had hoped for a first idea which changes in the new versions could lead to such a problem.
But I will try my best to create a StackBlitz reproduction.

**Comment by aleclarson** [comment]
The content of your `tsconfig.json` files might hint at the problem.

The only change in 4.3 is upgrading `tsconfck` (the package used for finding and parsing tsconfig files) by a major version.

**Comment by andersonba** [comment]
Same issue +1

**Comment by andersonba** [comment]
I'm experiencing the same issue as well

**Comment by Yberion** [comment]
Hello @aleclarson, I'm facing the same issue with Analogjs & Angular.

Here is a reproduction:

- https://github.com/Yberion/analog-nx/tree/vitest-tsconfig-4-3-0
- `npm i`
- `npm run test:app`

With `4.3.0` I'm facing this issue:

```sh
> nx run analog-nx:test

failed to load config from E:\Git\analog-nx\apps\analog-nx\vite.config.ts
 >  NX   No "exports" main defined in E:\Git\analog-nx\node_modules\tsconfck\package.json
```

With `4.3.1` nothing happen

![image](https://github.com/aleclarson/vite-tsconfig-paths/assets/4186385/f1b9fd65-686a-41a8-99a2-16c84af3c9b1)

Downgrading to `4.2.3` works for now.

**Comment by dominikg** [comment]
this seems to happen if you have an indirect circles. tsconfck detects circles in "extends", but a  circle thats between references and extends  can stay undetected and then it can lock up with 2 Promises waiting for each other if you try to parse a 2 of them at the same time.

This is bound to happen here https://github.com/aleclarson/vite-tsconfig-paths/blob/ca4e2aa7f3be36facbce0305cc145f9d5f9693d1/src/index.ts#L96  if all files in question are named `tsconfig.json`

but can also happen if many files are processed at the same time. I'll have to see how to guard against this in tsconfck.

A workaround in get-tsconfig-paths could be to serialize the parse of found configs for the time being, affected users could try to not have references point at files that they also extend.

circle:
```sh
/src/tsconfig.json # extends ../tsconfig.json
tsconfig.json  # references src/tsconfig.json
```

no circle:
```sh
/src/tsconfig.json # extends ../tsconfig.base.json
tsconfig.json  # references src/tsconfig.json , nothing but references in it
tsconfig.base.json  # all the base config values
```

**Comment by aleclarson** [comment]
If anyone wants to implement a workaround in this plugin, PR welcome.

Otherwise, use the config solution recommended by Dominik above, and wait for a fix to be added to `tsconfck` (a package used by this plugin for tsconfig parsing).

**Comment by bitschubse** [comment]
@aleclarson: Problem seems to be fixed with 4.3.2 without any changes on our side. Should I close the issue?
